src/systemd/sd-dhcp-server.h

```diff
@@ -31,7 +31,12 @@ _SD_BEGIN_DECLARATIONS;
 
 typedef struct sd_dhcp_server sd_dhcp_server;
 
-int sd_dhcp_server_new(sd_dhcp_server **ret, int ifindex);
+typedef struct sd_dhcp_server_cb {
+        void (*on_leases_changed)(sd_dhcp_server *s, void *data);
+} sd_dhcp_server_cb;
```

Comment:
Did you give any thought as to how this can be extended in the future? You're passing just a raw pointer to a structure, so there is no possibility to grow the structure in the future without breaking ABI. The API needs to be such that this can be extended in the future without recompiling all users.

This is a pretty big change. Please split this in two parts: the first should be the addition of the callback to the sd_dhcp_server and the second using of this new functionality in networkd. Please describe in the commit message why you are doing this, how the information is exposed over dbus and why, etc.

Reply:
What do you mean about breaking ABI? Isn't this a private structure?

Yes, I can split things further. I wanted to have some feedback before that.

So this added callback isn't doing much, it's just telling the networkd-dhcp-server-bus when the lease changes. But in the future, we may export more properties. So that's why I chose to have a struct with callbacks.

Reply:
It isn't exposed, but the plan is to do it at some point, and we're building an API that can be exposed. This means that anything that will be exposed:
- everything is namespaceder "sd_"
- those functions do more verification of arguments and use `assert_return` not `assert` on invalid arguments
- the api must be done in a way that backwards compat can be maintained in future changes

In this particular case, if you pass in the size of sd_dhcp_server_cb structure, if we add more fields to that structure, we can distinguish which version of the structure was used by the caller. (There are other options, but this one seems the simplest.)

Reply:
@poettering's approach seems better, so my comments here can be ignored.

Reply:
next update will be following poettering suggestion. thanks for your comments